src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,414 @@
+    if (!expr_match(pattern2, second, matches)) {
+        return {};
+    }
+    base += std::move(matches[0]);
+    Expr r_stride = std::move(matches[1]);
+
+    if (sub) {
+        Expr adj = sub->b;
+        const Broadcast *bcast = adj.as<Broadcast>();
+
+        if (!bcast) {
+            return {};
+        }
+
+        if (bcast->lanes != b1->lanes * r1->lanes) {
+            return {};
+        }
+
+        base -= bcast->value;
+    }
+
+    return {true, base, {x_stride, 0, r_stride}, {x_tile, y_tile, r_tile}};
+}
+
+struct NewMatmul {
+    bool result = false;
+    Stmt stmt;
+    int tile_x;
+    int tile_y;
+    int tile_r;
+};
+
+NewMatmul convert_to_matmul(const Store *op, const string &new_name, AMXOpType op_type) {
+    // m[ramp(0, 1, S)] = VectorAdd(lhs[{XYR tile}] * xX(rhs[{YR tile}])) + m[ramp(0, 1, S)]
+    const auto wild_i8x = Variable::make(Int(8, 0), "*");
+    const auto wild_u8x = Variable::make(UInt(8, 0), "*");
+    const auto wild_bf16x = Variable::make(BFloat(16, 0), "*");
+    const auto wild_f32x = Variable::make(Float(32, 0), "*");
+
+    vector<Expr> matches;
+    if (op_type == AMXOpType::Int8) {
+        const auto pattern1 = wild_i32x + wild_i32x;
+        if (!expr_match(pattern1, op->value, matches)) {
+            return {};
+        }
+    } else {  // AMXOpType::Bf16
+        const auto pattern1 = wild_f32x + wild_f32x;
+        if (!expr_match(pattern1, op->value, matches)) {
+            return {};
+        }
+    }
+
+    const auto *reduce = matches[0].as<VectorReduce>();
+    const auto *load = matches[1].as<Load>();
+    if (!reduce || reduce->op != VectorReduce::Add) {
+        return {};
+    }
+    if (!load || load->name != op->name || !equal(load->index, op->index)) {
+        return {};
+    }
+
+    if (op_type == AMXOpType::Int8) {
+        auto pattern2 = cast(Int(32, 0), cast(Int(32, 0), wild_i8x) * wild_i32x);
+        auto pattern2_unsigned = cast(Int(32, 0), cast(Int(32, 0), wild_u8x) * wild_i32x);
+
+        if (!(expr_match(pattern2, reduce->value, matches) || expr_match(pattern2_unsigned, reduce->value, matches))) {
+            return {};
+        }
+    } else {
+        auto pattern2 = cast(Float(32, 0), cast(Float(32, 0), wild_bf16x) * wild_f32x);
+
+        if (!expr_match(pattern2, reduce->value, matches)) {
+            return {};
+        }
+    }
+
+    const auto *lhs_load = matches[0].as<Load>();
+    const auto *rhs_broadcast = matches[1].as<Broadcast>();
+    if (!lhs_load || !rhs_broadcast) {
+        return {};
+    }
+    const auto *rhs_cast = rhs_broadcast->value.as<Cast>();
+    if (rhs_cast) {
+        if (op_type == AMXOpType::Int8) {
+            if (!(rhs_cast->value.type().element_of() == Int(8) || rhs_cast->value.type().element_of() == UInt(8))) {
+                user_assert(false) << "Expected rhs cast of i8/u8";
+            }
+        } else  // AMXOpType::Bf16
+        {
```

Comment:
Closing brace should be on previous line. (I'm surprised clang-format didn't catch this.)